### PR TITLE
Add feature tool: convert added feature into the right type consideri…

### PR DESCRIPTION
**Pull request by mhugent**
…ng captured curve type, layer geometry type and provider capabilities

### CODE REVIEW

File: src/app/qgsmaptooladdfeature.cpp [hunk 1]
```diff
@@ -339,3 +312,62 @@ void QgsMapToolAddFeature::cadCanvasReleaseEvent( QgsMapMouseEvent* e )
     }
   }
 }
+
+QgsAbstractGeometryV2* QgsMapToolAddFeature::outputGeometry( const QgsCurveV2* c, const QgsVectorLayer* v ) const
+{
+  if ( !c || !v )
+  {
+    return 0;
```
Comment (m-kuhn) [inline src/app/qgsmaptooladdfeature.cpp]: Should be `nullptr`

**Comment by m-kuhn** [comment]
Hi @mhugent

Thanks for the pull request.

* Is it the job of the tool to set the geometry type or should this be done just before sending it to the backend (or when it enters the edit buffer)?
* If this needs to be done in the map tool, shouldn't it be done in a superclass so other maptools like editing the geometry and vertices do the same?

What's the driver behind this request? Does it fix a bug or improve UX?

**Comment by mhugent** [comment]
Hi @m-kuhn

>What's the driver behind this request? Does it fix a bug or improve UX?

If fixes a pretty serious bug when working with curves in practice. Say you are adding a feature to a postgres table with type curve polygon. If the  added polygon has straight segments only, the tool makes a polygon and postgres complains once the feature is written back to the db.

>Is it the job of the tool to set the geometry type or should this be done just before sending it to the >backend (or when it enters the edit buffer)?
 
The situation in the map tools is a bit messy. The tools need to do a first conversion already before the edit buffer (e.g. convert line to polygon ring, convert to multitype) which varies slightly between tools. 

Don't know if altering before sending works in all cases, but it is worth a try.

**Comment by nyalldawson** [comment]
@mhugent can this be closed now? It's superseded, right?

**Pull request closed without merging**